nexus/reconfigurator/execution/src/datasets.rs

```diff
@@ -4,97 +4,214 @@
+use omicron_uuid_kinds::DatasetUuid;
 use omicron_uuid_kinds::GenericUuid;
-use omicron_uuid_kinds::OmicronZoneUuid;
+use omicron_uuid_kinds::SledUuid;
 use slog::info;
+use slog::o;
 use slog::warn;
-use std::collections::BTreeSet;
+use std::collections::BTreeMap;
+
+/// Idempotently ensures that the specified datasets are deployed to the
+/// corresponding sleds
+pub(crate) async fn deploy_datasets(
+    opctx: &OpContext,
+    sleds_by_id: &BTreeMap<SledUuid, Sled>,
+    sled_configs: &BTreeMap<SledUuid, BlueprintDatasetsConfig>,
+) -> Result<(), Vec<anyhow::Error>> {
+    let errors: Vec<_> = stream::iter(sled_configs)
+        .filter_map(|(sled_id, config)| async move {
+            let log = opctx.log.new(o!(
+                "sled_id" => sled_id.to_string(),
+                "generation" => config.generation.to_string(),
+            ));
+
+            let db_sled = match sleds_by_id.get(&sled_id) {
+                Some(sled) => sled,
+                None => {
+                    let err = anyhow!("sled not found in db list: {}", sled_id);
+                    warn!(log, "{err:#}");
+                    return Some(err);
+                }
+            };
+
+            let client = nexus_networking::sled_client_from_address(
+                sled_id.into_untyped_uuid(),
+                db_sled.sled_agent_address,
+                &log,
+            );
+
+            let config: DatasetsConfig = config.clone().into();
+            let result =
+                client.datasets_put(&config).await.with_context(
+                    || format!("Failed to put {config:#?} to sled {sled_id}"),
+                );
+            match result {
+                Err(error) => {
+                    warn!(log, "{error:#}");
+                    Some(error)
+                }
+                Ok(result) => {
+                    let (errs, successes): (Vec<_>, Vec<_>) = result
+                        .into_inner()
+                        .status
+                        .into_iter()
+                        .partition(|status| status.err.is_some());
 
-/// For each zone in `all_omicron_zones` that has an associated durable dataset,
-/// ensure that a corresponding dataset record exists in `datastore`.
+                    if !errs.is_empty() {
+                        warn!(
+                            log,
+                            "Failed to deploy datasets for sled agent";
+                            "successfully configured datasets" => successes.len(),
+                            "failed dataset configurations" => errs.len(),
+                        );
+                        for err in &errs {
+                            warn!(log, "{err:?}");
+                        }
+                        return Some(anyhow!(
+                            "failure deploying datasets: {:?}",
+                            errs
+                        ));
+                    }
+
+                    info!(
```

Comment:
My preference is to actually not use `else {` blocks, but to have a series of `if {` statements with early returns.

In an extreme case, I'd rather write:

```rust
if err1 {
  return;
}
if err2 {
  return;
}
do_something();
```

Instead of the more indented:

```rust
if err1 {
  return;
} else {
  if err2 {
    return;
  } else {
    do_something();
  }
}
```

I think my code is *currently* doing the first pattern, and adding an `else` makes it use the second pattern.